File: moveit_experimental/jog_arm/src/jog_arm/xbox_to_twist.py
#!/usr/bin/env python

# Take joystick cmds. Republish them as TwistStamped for
# e.g. jogging.

# RB: +x, LB: -x
# L on L dpad: +x
# Up on L dpad: +z

# R on R stick: +Rx
# Up on R stick: +Ry
# B: +Rz, A: -Rz

import rospy
from geometry_msgs.msg import TwistStamped
from sensor_msgs.msg import Joy

class joy_to_twist:

    def __init__(self):
        self.pub = rospy.Publisher('jog_arm_server/delta_jog_cmds', TwistStamped, queue_size=1)

    # Convert to TwistStamped and republish
    def callback(self, joy):

        ts = TwistStamped()

        ts.header.stamp = rospy.Time.now()

        # These buttons are binary
        ts.twist.linear.x = -joy.buttons[4] + joy.buttons[5]
        # Double buttons
        ts.twist.linear.y = joy.axes[0]
        ts.twist.linear.z = joy.axes[1]

        ts.twist.angular.x = -joy.axes[3]
        ts.twist.angular.y = joy.axes[4]
        # These buttons are binary
        ts.twist.angular.z = -joy.buttons[0] + joy.buttons[1]

        self.pub.publish(ts)
        
    def republish(self):
        rospy.Subscriber("joy", Joy, self.callback)

        rospy.spin()

if __name__ == '__main__':
    rospy.init_node('joy_to_twist', anonymous=True)
    republisher = joy_to_twist()
    republisher.republish()

File: moveit_experimental/jog_arm/src/jog_arm/xbox_to_twist.cpp
#include "geometry_msgs/TwistStamped.h"
#include "ros/ros.h"
#include "sensor_msgs/Joy.h"

namespace to_twist
{
class xboxToTwist
{
public:
  xboxToTwist() : spinner_(2)
  {
    joy_sub_ = n_.subscribe("joy", 1, &xboxToTwist::joyCallback, this);
    twist_pub_ = n_.advertise<geometry_msgs::TwistStamped>("jog_arm_server/delta_jog_cmds", 1);

    spinner_.start();
    ros::waitForShutdown();
  };

private:
  ros::NodeHandle n_;
  ros::Subscriber joy_sub_;
  ros::Publisher twist_pub_;
  ros::AsyncSpinner spinner_;

  // Convert incoming joy commands to TwistStamped commands for jogging
  void joyCallback(const sensor_msgs::Joy::ConstPtr& msg)
  {
    geometry_msgs::TwistStamped t_s;
    t_s.header.stamp = ros::Time::now();

    // This button is binary
    t_s.twist.linear.x = -msg->buttons[4] + msg->buttons[5];
    // Double buttons
    t_s.twist.linear.y = msg->axes[0];
    t_s.twist.linear.z = msg->axes[1];

    t_s.twist.angular.x = -msg->axes[3];
    t_s.twist.angular.y = msg->axes[4];
    // A binary button
    t_s.twist.angular.z = -msg->buttons[0] + msg->buttons[1];

    twist_pub_.publish(t_s);
  }
};
}  // end to_twist namespace

int main(int argc, char** argv)
{
  ros::init(argc, argv, "xbox_to_twist");

  to_twist::xboxToTwist to_twist;

  return 0;
}
